src/ffi/mod.rs

```diff
@@ -85,12 +119,604 @@ lazy_static::lazy_static! {
+        }
+        let serialized_event_fn = format!("{:?}", events);
+        let ret = Self::new_common(&features, events, None);
+        Self::log_entry(features, serialized_event_fn, &ret);
+        ret
+    }
+
+    /// Create new telio library instance
+    /// # Parameters
+    /// - `events`:     Events callback
+    /// - `features`:   JSON string of enabled features
+    /// - `protect`:    Callback executed after exit-node connect (for VpnService::protectFromVpn())
+    pub fn new_with_protect(
+        features: Features,
+        events: Box<dyn TelioEventCb>,
+        protect: Box<dyn TelioProtectCb>,
+    ) -> FFIResult<Self> {
+        let serialized_event_fn = format!("{:?}", events);
+        let ret = Self::new_common(&features, events, Some(protect));
+        Self::log_entry(features, serialized_event_fn, &ret);
+        ret
+    }
+
+    fn new_common(
+        features: &Features,
+        events: Box<dyn TelioEventCb>,
+        #[allow(unused)] protect_cb: Option<Box<dyn TelioProtectCb>>,
+    ) -> FFIResult<Self> {
+        let events = Arc::new(events);
+        let event_dispatcher = move |e: Box<Event>| {
+            let payload = e
+                .to_json()
+                .unwrap_or_else(|_| String::from("event_to_json error"));
+            events.event(payload);
+        };
+
+        let panic_event_dispatcher = event_dispatcher.clone();
+        PANIC_HOOK.call_once(|| {
+            let events = panic_event_dispatcher;
+            panic::set_hook(Box::new(move |info| {
+                // We need it on the logs as well ...
+                error!("{}", info);
+
+                let err = {
+                    let message = {
+                        if let Some(msg) = info.payload().downcast_ref::<String>() {
+                            msg.clone()
+                        } else if let Some(msg) = info.payload().downcast_ref::<&str>() {
+                            msg.to_string()
+                        } else {
+                            DEFAULT_PANIC_MSG.to_string()
+                        }
+                    };
+                    anyhow::Error::from(panic_handling::Panic { message })
+                };
+
+                // Updating LAST_ERROR.
+                // NOTE: this "could" duplicate updating error, if the error happens on ffi call stack as well ...
+                error_handling::update_last_error(err);
+
+                // Send this "cry for help" to whoever is on the upper side
+                let e = Box::new(
+                    Event::new::<Error>()
+                        .set(ErrorCode::Unknown)
+                        .set(ErrorLevel::Critical)
+                        .set(format!("{}", info)),
+                );
+
+                telio_log_debug!("call_once: {:?}", e);
+                events(e);
+            }));
+        });
+
+        // TODO: Update windows ffi to take in void*, for protect
+        #[cfg(not(target_os = "android"))]
```

Comment:
Do we really need this to be gatted under android. I think the option is universal enough, and in the future someone needs to customize this, I see no problem with that.

Reply:
I was following the old logic here, but I'm happy to not conditionally compile this

Reply:
It seems like file descriptors are u64 on windows but i32 on all other platforms. I can either try to dig into the code and add a conversion wherever windows is calling it, or a simpler option, exclude the protect closure on windows. Opinion?

Reply:
so let's this this for unix only